ROADMAP.md

```diff
@@ -1,45 +1,40 @@
+    - We are finishing the spec at the moment which will then be published as an RFC for review and feedback.
+    - Once we're ready to begin implementation those details and development phases will be shared here.
 
-- **Swift Codegen Rewrite, Continued**: As outlined in much greater detail in the [RFC issue](https://github.com/apollographql/apollo-ios/issues/939), we're rewriting our code generation. This has taken ~~somewhat~~ ~~a lot~~ colossally longer than expected, but the following phases are still in progress:
-    
-    - **Start generating code with Swift instead of Typescript**. This will allow much easier community contribution to code generation, and allow us to take on a bunch of improvements like `Hashable`, `Equatable`, and potentially `Identifiable` without having to fight with Typescript to do it.
-    - **Add immutable caching to new generated code**. Caching is currently heavily tied into our existing parsing mechanism. We're going to separate that out in two phases: The first will allow caching that *cannot* be changed by the consumer.
-    - **Add mutable caching to new generated code**. This is the final stage of updating caching: Allowing caching that *can* be changed by the consumer.
-    - **Remove Old Codegen**. Once all this is built, older codegen will be deprecated. 
-
-    You can follow this effort through the [Swift Codegen GitHub project](https://github.com/apollographql/apollo-ios/projects/2).
-    
-- **Cache Improvements**: We've had a ton of requests for this and we've got some extra muscle now, so we're going to be working on many requests. First up, some features inspired by Apollo Client 3 for the web: 
-    - **Better pagination support**. Better support for caching and updating paginated lists of objects. 
-    - **Reducing over-normalization**. Only separating out results into individual records when something that can identify them is present
-    - **Codable Support**: Which will allow easier moving between any storage method and objects, along with working better with updated codegen.
-
-- **General bug-bashing**: There are always a few outstanding general issues and small feature requests I'd like to address, and we'll be dealing with any new issues as they come up.
+- **Networking Stack Improvements**: The goal is to simplify and stabilise the networking stack.
+    - The [updated network stack](https://github.com/apollographql/apollo-ios/issues/1340) solved a number of long standing issues with the old barebones NetworkTransport but still has limitations. Switching to a composable request chain would enable building a pipeline out of reusable and customizable parts, giving users full control over the steps that are invoked to satisfy requests. This pattern has proven really useful for the JavaScript client (Apollo Link), and has been used succesfully in a number of other projects (like [OkHttp](https://square.github.io/okhttp/)).
```

Comment:
I think this captures the gist of what we discussed earlier and this description exists elsewhere, albeit with minor changes.

Reply:
We've got **a** composable request chain at this point with the `RequestChainNetworkTransport` and interceptors, it's just that setting it up is Too Fucking Complicated™, and it doesn't line up with the Apollo-Link setup as much as with at least part of the `OKHTTPInterceptor` setup used on Android (it's not completely aligned with that either, but it's a hell of a lot closer).

Basically, the main difference is that in our current setup, we're expecting that each individual interceptor doesn't need or want to know anything about what interceptors have run before or will run after. Apollo-Link on the other hand allows each link to decide what the next link is (eg, if it's a subscription use a websocket link, if it's a mutation or a query use a normal http link). 

That's a major philosophical difference that when I was first working on this I really didn't like, but after seeing the shortcomings of the direction I went I can see the case for a lot better now.

Reply:
Android 2 cents: `3.x` has both:

* [HttpInterceptor](https://github.com/apollographql/apollo-android/blob/3e44bea11bcff2bdfa27c18c33457e82da6aedd4/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpInterceptor.kt#L10) for HTTP requests
* [ApolloInterceptor](https://github.com/apollographql/apollo-android/blob/3e44bea11bcff2bdfa27c18c33457e82da6aedd4/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/ApolloInterceptor.kt#L12) for GraphQL request

Both of them are "OkHttp-like" in that they work in isolation and do not know anything about what's upstream and downstream.

They can communicate through an opaque `ExecutionContext` for `ApolloInterceptor` and HTTP headers for `HttpInterceptors`

I'd expect most users to use. `HttpInterceptor`. Authorization and Logging are the first examples that comes to mind. `ApolloInterceptors` are most likely used internaly for cache/apqs and significantly more complex to implement although I think it shouldn't be required very often.

Reply:
Fair points. Here's another attempt - d87c3318.